### DISCUSSION TITLE
Performance Issue comparing to the blenderbim addon

### DISCUSSION THREAD

**Discussion by yenshin** [IfcOpenShell/IfcOpenShell#2232] [Q&A]
Hello,

I'm working on a c++ project using IfcOpenShell lib.
I aim to load the model 2102_cira.zip, in attachement, as fast as blenderbim addon, without success.
My code load this model in 22 seconds, blenderbim in about 4 to 7 seconds.

To fix my problem i try different things:

- set ENABLE_BUILD_OPTIMIZATIONS to ON
- different git branch v0.7.0, revert-2142-refactor_georeference_module, blenderbim-220606 (containing ifcopenshell lib) 
- use different open collada version (OCCT OCE) (i don't test this on every version i tryed)

So now i have some question:

- would i be missing a compilation parameter for IOS lib that could bring performance boost ?
- would i be missing in my project a compilation parameter for performance boost ?
- should i need to change the way I iterate through an ifc file ?
- should i try minsys2 instead of visual ?
- any other idea ?

Thanks you for your time I wish you a nice day

[IOScompiltool.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/8871441/IOScompiltool.zip) compilation protocol
[libtest.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/8871445/libtest.zip) minimum code to load an ifc file
[2102_cira.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/8871447/2102_cira.zip) original source from http://openifcmodel.cs.auckland.ac.nz/Model/Details/374 but down when i write this.

**Comment by aothms** [comment]
This file has a lot of (indexed) explicit polygonal meshes.

Open Cascade is not very efficient in handling these. The BRep model has a lot of overhead for this particular case (edge curves, face surfaces, ...).

BlenderBIM bypasses Open Cascade in these cases. You can read more about it here: #841 

It is on the roadmap to generically handle this in IfcOpenShell by having alternative geometry implementations, but having this reliable and production ready is at least one year out.

**Reply by Moult** [reply]
I don't think that's it. The BlenderBIM Add-on actually doesn't bypass anything, the code was always commented out and only recently I put a toggle in the UI but by default it's still disabled.

A few things I can think of:

 - I see USE_WORLD_COORDINATES, that break instancing and is very, very slow from what I've experienced
 - Are you using multiprocessing in C++?
 - Make sure you are comparing non-cached imports in the BlenderBIM Add-on. I've recently disabled it by default due to stability (segfaults on large files) but in the stable build subsequent imports read from a HDF5 cache. That said I just tried a cold import and it took 4.7s on my machine. So it's probably not the major factor here.

**Comment by yenshin** [comment]
Thank you all for the answers.

So I tried `USE_WORLD_COORDINATES = false`, and my loading time was halved (from 24 s to 12). It's great.

`Are you using multiprocessing in C++?`
I can say that i use multicore with theses lines:

> int maximumThread(std::thread::hardware_concurrency());
> IfcGeom::Iterator context_iterator(settings, file, maximumThread);

`Make sure you are comparing non-cached imports in the BlenderBIM Add-on.`

I can't find where i can disable the cache import. 


So for now, blender is importing the model in less than 4.7s and my test program in 12.4s
It's great improvement, but I thinks if their is no by pass in blender, the loading time should be at least the same or lower (in the test program we do not use the data, c++ compiled etc ...) .


If I find something else I 'll share it.
Thanks again, and have a nice days.

**Reply by Moult** [reply]
Are you doing anything with the shapes you get from `context_iterator.get()`? Or are you just discarding the data immediately as shown in your minimum example?

Also note that in the BlenderBIM Add-on we use set_context_ids (https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/import_ifc.py#L300) this limits the iterator to process only the representations that are part of that context (most commonly, Body). Otherwise, especially in models with lots of contexts, you will end up easily doubling or tripling the work the iterator needs to do which takes a huge amount of time.

The latest BlenderBIM Add-on builds have a GUI option where you can uncheck "cache" in the advanced settings when loading an IFC. If you don't have access to this, just edit this file and delete any lines that do `iterator.set_cache(...)` like this: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/import_ifc.py#L695 - there should be two of them (one for 3D one for 2D).

**Comment by yenshin** [comment]
`The latest BlenderBIM Add-on builds have a GUI option where you can uncheck "cache" in the advanced settings when loading an IFC. If you don't have access to this, just edit this file and delete any lines that do iterator.set_cache(...) like this: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/import_ifc.py#L695 - there should be two of them (one for 3D one for 2D).`

So I tried this, performance improvement is not really relevant, 5 seconds instead of 4. I'm not sure if performance loss is from this or because of other software, or measurement. 

`Are you doing anything with the shapes you get from context_iterator.get()? Or are you just discarding the data immediately as shown in your minimum example?`

I'm working on ifc model viewer. So with this I do the following:

- Get the triangulate model, like this: 
  - IfcGeom::TriangulationElement ***elem**(static_cast<IfcGeom::TriangulationElement *>(context_iterator.get());
- Generate and show an opengl version of the model with vertex and indices.
- Show Ifc hierarchy (IfcProject -> IfcSite -> IfcBuilding ... TriangulatedModel (IfcWall, IfcStair, etc...)
- Show **elem** metadata (Id, Ifc class name ...)

I replaced the use of` USE_WORLD_COORDS = true` by a vertex matrix multiplication (using [glm lib](https://github.com/g-truc/glm)).
Timings are; 

- 12.5 seconds with only `context_iterator.get()`
- 14 seconds (1.5 second longer) with parsing (indexes, vertex and matrix multiplication), and print on screen with Opengl

Here is my code if you need it: [https://pastebin.com/EBc3Yfcs](https://pastebin.com/EBc3Yfcs)

`Also note that in the BlenderBIM Add-on we use set_context_ids`

I try some modification using context ids, without any improvement. May be I'm missing something ?
Here's my code: [https://pastebin.com/MvCGB2hc](https://pastebin.com/MvCGB2hc)

Should I made my parsing differently ? I look to [IfcParseExamples.cpp](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/examples/IfcParseExamples.cpp) but I would like to avoid specifying IfcSchema.

Again, thanks for your time, and I wish you a nice day. 


note: I try another file [Duplex_Plumbing_20121113.ifc](https://github.com/buildingSMART/Sample-Test-Files/blob/master/IFC%202x3/Duplex%20Apartment/Duplex_Plumbing_20121113.ifc):
- 9 seconds for BlenderBIM Add-on
- 15 seconds with only `context_iterator.get()`

**Comment by aothms** [comment]
I missed this bit:

> use different open collada version (OCCT OCE) (i don't test this on every version i tryed)

This would be the first thing to try then. It depends a bit on what version of OCE you took, but for quite a while they were seriously lagging behind the official OCCT. If it means you end up using the OCE version of OCCT 6.9.x you'd miss out on a lot of performance improvements in OCCT wrt multi argument boolean operations.

Looking at your code, maybe also try without `ELEMENT_HIERARCHY` It's not written with performance in mind, don't know the performance impact, but I doubt it's the cause.

If you compile your own IfcOpenShell, did you try replacing the _ifcopenshell_wrapper.pyd that comes out of your build with the one in blenderbim to rule out any compilation settings?
